url70ed135da7_ralfbiedert_jcores_tgz-pJ8-sandbox_SimpleFilterTestJ8/src/net/jcores/jre/cores/Core.java

```diff
@@ -0,0 +1,414 @@
+
+        // Request a CPU for each element we have (in case we have many, we only receive maxCPU, in case we have 
+        // very few, we don't block all CPUs.
+        final int NUM_THREADS = Math.min(toGo, this.commonCore.profileInformation().numCPUs); // cc.requestCPUs(toGo);
+        
+        // We use a safetey factor of 2 for the fork time (FIXME: Should investigate what's the best factor),
+        // also, we only spawn something if there is more than one element still to go.
+        if((estTime < 2 * profileInfo.forkTime && toGo > 1) || NUM_THREADS < 2) {
+            // Instantly release all CPUs when we go singlethreaded
+            // this.commonCore.releaseCPUs(NUM_THREADS);
+            
+            // In this case, we go single threaded
+            while(iterator.hasNext()) {
+                final int i = iterator.nextIndex();
+                iterator.next(); // We need to get the next() that the nextIndex increases. 
+                mapper.handle(i);
+            }
+            
+            return;
+        }
+
+
+        // TODO: Get proper value for step size (same problem, see below)
+        // TODO: Check size, if small, don't do all this setup in here ...
+        // NAH, even for two objects we can have a speed gain if the calls
+        // are very slow ...
+
+        
+        // Okay, in this case the loop was broken and we decided to go parallel. In that case
+        // setup the barrier and spawn threads for all our processors so that we process the array.
+        final CyclicBarrier barrier = new CyclicBarrier(NUM_THREADS + 1);
+        final AtomicInteger baseCount = new AtomicInteger();
+
+        final Runnable runner = new Runnable() {
+            public void run() {
+                int bc = baseCount.getAndIncrement() * STEP_SIZE;
+                int lower = Math.max(index.get(), bc);
+                
+
+                // Get new basecount for every pass ...
+                while (lower < size) {
+                    final int max = Math.min(Math.min(lower + STEP_SIZE, size), bc + STEP_SIZE);
+
+                    // Pass over all elements
+                    for (int i = lower; i < max; i++) {
+                        mapper.handle(i);
+                    }
+
+                    bc = baseCount.getAndIncrement() * STEP_SIZE;
+                    lower = bc;
+                }
+
+                // Signal finish
+                try {
+                    barrier.await();
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                } catch (BrokenBarrierException e) {
+                    e.printStackTrace();
+                }
+            }
+        };
+
+        // Execute all our runnables.
+        for(int i=0; i<NUM_THREADS; i++) {
+            cc.executor().getExecutor().execute(runner);
+        }
+
+        // Wait for all threads to finish ...
+        try {
+            barrier.await();
+        } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)